core/src/main/java/org/apache/calcite/rel/rules/ReduceExpressionsRule.java

```diff
@@ -270,6 +270,8 @@ private void reduceNotNullableFilter(
           } else {
             call.transformTo(createEmptyRelOrEquivalent(call, filter));
           }
+          // New plan is absolutely better than old plan.
+          call.getPlanner().setImportance(filter, 0.0);
```

Comment:
`setImportance(filter, 0.0)` is not suggested to be used, i kind of remember see https://issues.apache.org/jira/browse/CALCITE-2223

Reply:
Thanks for sharing, @danny0405 . IMO, if we already do it in this rule[1], it makes sense to do it to be consistent.

[1] https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/rel/rules/ReduceExpressionsRule.java#L216

Reply:
Usages of `setImportance` indeed require some care. I think, this change for example could lead to a `CannotPlanException` that could be avoided before. 

Assume that we enter the else branch and we create an empty `LogicalValues`. Then if the rule set does not contain a `PhysicalValues` we may end-up with a `CannotPlanException`. Without this change this wouldn't be the case since the old plan could still be matched by other rules and create a `PhysicalFilter` etc. I admit this is a contrived example but I wanted to highlight that setting importance explicitly might be risky.

Reply:
> Assume that we enter the else branch and we create an empty LogicalValues. Then if the rule set does not contain a PhysicalValues we may end-up with a CannotPlanException.

This is not valid. Because without this change, `ReduceExpressionsRule` can also create an empty `LogicalValues`[1]. I agree that we should pay more attention when we use `setImportance`. But as I mentioned before, what this change does is to be consistent with what the if branch does.


[1] https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/rel/rules/ReduceExpressionsRule.java#L199

Reply:
OK, I didn't see [1] before. 

Other than maybe we could refactor slightly the code to keep the `setImportance` in one place (in [2]); leaving the decision up to you.

Reply:
IMHO, setImportance should be used as long as the new plan expression is better than old one, means just replace it. Even if there can be exception during planning, it is the fault of VolcanoPlanner, not the fault of `setImportance`. We should fix the VolcanoPlanner if the exception happens.